### DISCUSSION TITLE
These scripts are written with Bash-specific syntax and thus require the Bash shell specifically.

### DISCUSSION THREAD

**Discussion by tteck** [tteck/Proxmox#2536] [Announcements]
Taking this into account, the scripts have been updated to include a check to verify if the default shell is set to Bash. If the default shell is not Bash, the script will exit.

**_This update aims to prevent difficult-to-troubleshoot issues that can arise from using non-Bash shells_** (_apologies to Zsh users_)**.**

**Comment by szymswiat** [comment]
To clarify: do we have to set bash as the default shell from now on? Executing e.g. `bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/docker.sh)"` is not enough? Use of `basename "$SHELL"` in https://github.com/tteck/Proxmox/commit/0b604d204a174e567477e38ef75c6e4f1e90a9a2 was intended?

**Reply by tteck** [reply]
If you haven't changed the default shell, you're good. However, using _Zsh_ may cause issues due to its slightly different syntax and behavior compared to _Bash_, particularly with constructs such as arrays, variable expansion, and command substitution.

The `basename` command is used to extract the filename from a given path. For example, if `$SHELL` contains `/bin/bash`, then `basename "$SHELL"` will output `bash`, similarly, if `$SHELL` contains `/usr/bin/zsh`, `basename "$SHELL"` will output `zsh`.

In the newly created `shell_check` function, if the default shell is not set to `bash`, the script will exit.
https://github.com/tteck/Proxmox/blob/9d9620249f8f501a6c969d6829f5d0ebec761a05/misc/build.func#L75-L83
Here is a recent example, https://github.com/tteck/Proxmox/issues/2440#issuecomment-1957723532, there are many more scattered throughout.

**Comment by tteck** [comment]
@szymswiat were you using Zsh as your shell while experiencing the issue in https://github.com/tteck/Proxmox/issues/2410?

**Reply by szymswiat** [reply]
@tteck my base shell is zsh however I have run it with `bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/omv.sh)"` which utilizes bash. I don't think checking $SHELL variable is a good approach since it contains path to login shell and it won't change even after entering bash. Please refer to https://unix.stackexchange.com/questions/522907/shell-environment-variable-still-points-to-zsh-after-using-bash. I think [checking BASH_VERSION](https://unix.stackexchange.com/a/522917) is more reliable. For now as a workaround I need to set `SHELL=/bin/bash bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/omv.sh)"` to let it work.

**Reply by tteck** [reply]
When you use `bash -c` to execute the script, you're indeed ensuring that the script is interpreted and executed with the Bash shell. _However_, if the script relies on **Bash-specific** features or behaviors, it may not work properly when executed with Zsh, even if it's invoked through `bash -c`. I've observed this in several reported issues.

So, the `$SHELL` variable will still point to `/bin/zsh`, even though the script is being executed with `bash -c`. This addresses the issue by disallowing the use of Zsh.

**Reply by lightmaster** [reply]
@szymswiat  Thanks for this. I had forgotten you could set a variable prior to running a command. Makes it a lot easier to keep using the scripts while this bizarre issues still persists.

**Reply by lightmaster** [reply]
@szymswiat A better workaround I found is to setup an alias in zsh, so that every time you run a command that starts with `bash` as it's first word, instead of actually running `bash foo bar`, you'll run `SHELL=/bin/bash bash foo bar`. This includes copy/pasting these scripts that start with `bash -c`, so it'll automatically prepend `SHELL=/bin/bash` in front of `bash -c`.

1. Add this to the end of your `~/.zshrc` (or create if it doesn't exist and add this to it)

```zsh
if [ -f ~/.aliases ] && [ -r ~/.aliases ]; then
        . ~/.aliases
fi
```

2. Create `~/.aliases` and add this to it

```zsh
alias bash="SHELL=/bin/bash bash"
```

3. Reload ZSH with the new `.zshrc`

```zsh
. ~/.zshrc
```

**Comment by tteck** [comment]
I've identified several issues that I believe stem from having zsh set as the default shell, and I suspect there may be many more.
https://github.com/tteck/Proxmox/issues/2440#issuecomment-1957723532
https://github.com/tteck/Proxmox/issues/2410#issue-2108885604
https://github.com/tteck/Proxmox/issues/2200#issuecomment-1875794748
https://github.com/tteck/Proxmox/discussions/2411#discussioncomment-8316830
https://github.com/tteck/Proxmox/issues/524#issuecomment-1255471993
https://github.com/tteck/Proxmox/issues/1698#issue-1852573406
